lib/chef/event_dispatch/dsl.rb

```diff
@@ -0,0 +1,54 @@
+#
+# Author:: Ranjib Dey (<ranjib@linux.com>)
+# Copyright:: Copyright (c) 2015 Ranjib Dey
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+require 'chef/event_dispatch/base'
+require 'chef/exceptions'
+require 'chef/config'
+
+class Chef
+  module EventDispatch
+    class DSL
+      # Adds a new event handler derived from base handler
+      # with user defined block against a chef event
+      #
+      # @return [Chef::EventDispatch::Base] a base handler object
+      def on(event_type, &block)
+        validate!(event_type)
+        handler = Chef::EventDispatch::Base.new
```

Comment:
The other way to go would be to create a class for each invocation of `Chef.event_handler`. For example,

```
def initialize(handler_class_name)
  @handler_class = Class.new(Chef::EventDispatch::Base)
  set_handler_class_const_in_sandbox_namespace(@handler_class)
  @handler_instance = @handler_class.new
end
```

Then the blocks you pass to the `on` method would just be the method bodies and instance variables and such should work.